### PR TITLE
Open PDF attachments in the viewer instead of an unconditional download

### LINKED ISSUE #6643: Open PDF-as-attachment in PDF.js
When a PDF contains other PDFs as attachments and the user clicks on a PDF-as-attachment, the attachment should be opened in PDF.js (either same tab or new tab) instead of being prompted to download.

PDF document: http://www.jasnapaka.com/bugs/aaa.pdf
OS: Windows 7 64 bit
pdf.js: Firefox 42 (32bit, release), PDF.js: ?

### PR DESCRIPTION
**Pull request by Rob--W**
To test:

1. Open the PDF from #6643.
2. Click on the attachment list.
3. Click on a PDF, and confirm that it opens in a new tab.

Tested with Firefox 51 (generic target), Chrome 56 (extension).

Fixes #6643
Relevant for #4282

### CODE REVIEW

File: web/pdf_attachment_viewer.js [hunk 1]
```diff
@@ -89,6 +89,34 @@ var PDFAttachmentViewer = (function PDFAttachmentViewerClosure() {
     /**
      * @private
      */
+    _bindPdfLink:
+        function PDFAttachmentViewer_bindPdfLink(button, content, filename) {
+      var blobUrl;
+      button.onclick = function() {
+        if (!blobUrl) {
+          blobUrl = pdfjsLib.createObjectURL(content, 'application/pdf') +
```
Comment (Snuffleupagus) [inline web/pdf_attachment_viewer.js]: Nit: Similar to e.g. https://github.com/mozilla/pdf.js/blob/a719b71e5957ee513ebf43a4ff703926c629b553/web/download_manager.js#L83-L84, you may want to add `pdfjsLib.PDFJS.disableCreateObjectURL` as the third parameter here!?

File: web/pdf_attachment_viewer.js [hunk 2]
```diff
@@ -89,6 +89,34 @@ var PDFAttachmentViewer = (function PDFAttachmentViewerClosure() {
...
+        if (typeof PDFJSDev === 'undefined' || PDFJSDev.test('GENERIC')) {
+          // The current URL is the viewer, let's use it and append the file.
+          viewerUrl = '?file=' + encodeURIComponent(blobUrl);
+        } else if (PDFJSDev.test('CHROME')) {
+          // eslint-disable-next-line no-undef
+          viewerUrl = chrome.runtime.getURL('/content/web/viewer.html') +
+            '?file=' + encodeURIComponent(blobUrl);
+        } else {
+          // Let Firefox's content handler catch the URL and display the PDF.
+          viewerUrl = blobUrl;
```
Comment (Snuffleupagus) [inline web/pdf_attachment_viewer.js]: In the Firefox addon, the `#` used above causes the viewer to try and goto a named destination equal to the `filename`. Most likely such a destination won't exist, but it'd be great to try and avoid this altogether.

**Review by timvandermeij — COMMENTED** [review]
This patch already looks good, but I would like to see the code with the existing comments addressed before approving this change. Thanks!

File: web/pdf_attachment_viewer.js (hunk 2 above)
Comment (Rob--W) [inline web/pdf_attachment_viewer.js]: Oh I see. It turns out that there are two ways to obtain the file name.

- `getFilenameFromUrl` (originally introduced in 83b6eaed1602ec3ddcbfe6633b943ecfb3da3e10)
- `getPDFFileNameFromURL` (84ae29c21d0bcdad09cb8ff71d53d4a149100fac from #3161)

I have updated the PR to use the latter.

File: web/pdf_attachment_viewer.js (hunk 1 above)
Comment (Rob--W) [inline web/pdf_attachment_viewer.js]: Done.

File: web/pdf_attachment_viewer.js [hunk 3]
```diff
@@ -89,6 +89,43 @@ var PDFAttachmentViewer = (function PDFAttachmentViewerClosure() {
...
+        if (typeof PDFJSDev === 'undefined' || PDFJSDev.test('GENERIC')) {
+          // The current URL is the viewer, let's use it and append the file.
+          viewerUrl = '?file=' + encodeURIComponent(blobUrl + '#' + filename);
+        } else if (PDFJSDev.test('CHROME')) {
+          // In the Chrome extension, the URL is rewritten using the history API
+          // in viewer.js, so an absolute URL must be generated.
+          // eslint-disable-next-line no-undef
+          viewerUrl = chrome.runtime.getURL('/content/web/viewer.html') +
+            '?file=' + encodeURIComponent(blobUrl + '#' + filename);
+        } else {
```
Comment (Snuffleupagus) [inline web/pdf_attachment_viewer.js]: Do we actually need to use such complicated logic below, since this code block really only ought to be relevant for `FIREFOX`/`MOZCENTRAL` builds, where `URL.createObjectURL` should always be supported anyway?

I.e. wouldn't it be sufficient to do something like the below instead, and just ignore the `pdfjsLib.PDFJS.disableCreateObjectURL === true` case; given that there's no way for a user of the addon/built-in version to disable `createObjectURL` short of editing the source code.
```javascript
} else if (PDFJSDev.test('FIREFOX || MOZCENTRAL')) {
  // Let Firefox's content handler catch the URL and display the PDF.
  viewerUrl = blobUrl + '?' + encodeURIComponent(filename);
}
```

Comment (Rob--W) [inline web/pdf_attachment_viewer.js]: I'm fine with that suggestion. Is `PDFJS.disableCreateObjectURL` mainly intended for use in websites?

Comment (Snuffleupagus) [inline web/pdf_attachment_viewer.js]: > Is PDFJS.disableCreateObjectURL mainly intended for use in websites?

It's my understanding that its only reason for existence is for compatibility purposes (with IE, refer to PR #4103), see https://github.com/mozilla/pdf.js/blob/master/web/compatibility.js#L458-L465.

Comment (Rob--W) [inline web/pdf_attachment_viewer.js]: I added a separate commit that removes the `disableCreateObjectURL` check.

**Comment by timvandermeij** [comment]
Could you squash the commits a bit? Having six commits seems a bit much for this relatively small change. Let's try to reduce it to two or three commits if possible.

**Comment by Rob--W** [comment]
Each individual commit serves their own purpose, so I'd like to keep them separate. If the commit descriptions are not clear I am willing to elaborate more.

File: web/pdf_attachment_viewer.js (hunk 3 above)
Comment (Snuffleupagus) [inline web/pdf_attachment_viewer.js]: > I added a separate commit that removes the disableCreateObjectURL check.

To me, it doesn't look entirely necessary for this to be a separate commit. Hence I'd suggest just folding into the previous commit, which also help slightly with https://github.com/mozilla/pdf.js/pull/8024#issuecomment-278182396.

Comment (Rob--W) [inline web/pdf_attachment_viewer.js]: Done. I initially kept them separate in case someone ever has a need for disabling `createObjectURL` in Firefox (and wants to be able to revert the commit), but now I have squashed the two commits with an extra comment and updated commit message.

**Review by timvandermeij — APPROVED** [review]

**Comment by timvandermeij** [comment]
Thank you for the patch!
